File: TargetedDosColisionsAndRNGAngle/web/lib/fabric.ts
// TS mirror of testbed/topology/fabric.py (plan-8 D8-parity). Tier A only
// needs to *demonstrate* the mechanism, not reproduce Python bit-for-bit
// (same stance as lib/salt.ts) -- but the fat-tree shape (k=4, 20 switches,
// 16 hosts) and the routing algorithm (hashed upward, deterministic downward)
// match testbed/topology/fabric.py exactly, and `ecmpLink` is P2's vendored
// hash unchanged.
import { ecmpLink, type FiveTuple } from "./ecmpHash";

export const FATTREE_K = 6;

// Web-viz-only super-spine: a small WAN-gateway tier above the core, modelling
// the real datacenter north-south egress path (core -> border/WAN routers).
// Not part of the Python fabric.py mirror -- it carries only the egress flow
// set (see wanFlowSet/egressRouteNodes) and adds a fifth vertical tier.
export const GATEWAY_COUNT = 2;

// Hosts per edge switch is decoupled from k here (web-viz only): a real k=6
// fat-tree has k/2=3 hosts/edge (54 hosts), but 2/edge (36 hosts) keeps the
// bottom row readable while leaving the switch fabric full-size.
export const HOSTS_PER_EDGE = 2;

export interface Fabric {
  k: number;
  half: number; // agg_per_pod = edge_per_pod = hosts_per_edge = core_per_group
  edgeSwitches: string[];
  aggSwitches: string[];
  coreSwitches: string[];
  gateways: string[];
  allSwitches: string[];
  hosts: string[];
  ipToHost: Record<string, string>;
  hostToIp: Record<string, string>;
  hostEdge: Record<string, string>;
  edgePod: Record<string, number>;
  aggPod: Record<string, number>;
  aggLocalIndex: Record<string, number>;
  linkIds: string[];
}

function linkId(a: string, b: string): string {
  return [a, b].sort().join("-");
}

export function buildFattree(k: number = FATTREE_K, hostsPerEdge: number = HOSTS_PER_EDGE): Fabric {
  const half = k / 2;
  const edgeSwitches: string[] = [];
  const aggSwitches: string[] = [];
  const coreSwitches: string[] = Array.from({ length: half * half }, (_, c) => `core${c}`);
  const gateways: string[] = Array.from({ length: GATEWAY_COUNT }, (_, g) => `gw${g}`);
  const allSwitches: string[] = [...coreSwitches];
  const hosts: string[] = [];
  const ipToHost: Record<string, string> = {};
  const hostToIp: Record<string, string> = {};
  const hostEdge: Record<string, string> = {};
  const edgePod: Record<string, number> = {};
  const aggPod: Record<string, number> = {};
  const aggLocalIndex: Record<string, number> = {};
  const linkIds: string[] = [];

  for (let pod = 0; pod < k; pod++) {
    for (let agg = 0; agg < half; agg++) {
      const aggId = `pod${pod}agg${agg}`;
      aggSwitches.push(aggId);
      allSwitches.push(aggId);
      aggPod[aggId] = pod;
      aggLocalIndex[aggId] = agg;
    }
    for (let edge = 0; edge < half; edge++) {
      const edgeId = `pod${pod}edge${edge}`;
      edgeSwitches.push(edgeId);
      allSwitches.push(edgeId);
      edgePod[edgeId] = pod;
      for (let agg = 0; agg < half; agg++) {
        linkIds.push(linkId(edgeId, `pod${pod}agg${agg}`));
      }
      for (let host = 0; host < hostsPerEdge; host++) {
        const hostId = `pod${pod}edge${edge}host${host}`;
        const ip = `10.${pod}.${edge}.${host + 1}`;
        hosts.push(hostId);
        ipToHost[ip] = hostId;
        hostToIp[hostId] = ip;
        hostEdge[hostId] = edgeId;
      }
    }
    for (let agg = 0; agg < half; agg++) {
      const aggId = `pod${pod}agg${agg}`;
      const groupStart = agg * half;
      for (let core = groupStart; core < groupStart + half; core++) {
        linkIds.push(linkId(aggId, `core${core}`));
      }
    }
  }

  // Super-spine uplinks: every core connects to every WAN gateway.
  for (const gw of gateways) {
    for (const core of coreSwitches) linkIds.push(linkId(core, gw));
  }

  return {
    k,
    half,
    edgeSwitches,
    aggSwitches,
    coreSwitches,
    gateways,
    allSwitches,
    hosts,
    ipToHost,
    hostToIp,
    hostEdge,
    edgePod,
    aggPod,
    aggLocalIndex,
    linkIds,
  };
}

/** Per-switch salt for every switch AND WAN gateway. `mintSalt` is supplied by
 * the caller (lib/salt.ts) so this module stays salt-source agnostic:
 *   - weak-prng: `mintSalt` cycles a tiny fixed pool (short-period LCG), so many
 *     switches share the same few salts -> correlated hash choices across tiers
 *     -> polarization spread over the whole fabric as hot/cold imbalance.
 *   - csprng/qrng: every call returns an independent draw -> uncorrelated. */
export function fabricSalts(kind: "weak-prng" | "csprng" | "qrng", fabric: Fabric, mintSalt: () => string): Record<string, string> {
  const salts: Record<string, string> = {};
  for (const switchId of [...fabric.allSwitches, ...fabric.gateways]) salts[switchId] = mintSalt();
  return salts;
}

/** Walk src-edge -> (hashed) agg -> [same pod: straight to dst-edge | different
 * pod: (hashed) core -> (deterministic) dst-agg -> (deterministic) dst-edge].
 * Mirrors testbed/topology/fabric.py `route()` exactly. */
export async function route(fabric: Fabric, salts: Record<string, string>, fiveTuple: FiveTuple): Promise<string[]> {
  const srcEdge = fabric.hostEdge[fabric.ipToHost[fiveTuple.src_ip]];
  const dstEdge = fabric.hostEdge[fabric.ipToHost[fiveTuple.dst_ip]];
  if (srcEdge === dstEdge) return [];

  const srcPod = fabric.edgePod[srcEdge];
  const dstPod = fabric.edgePod[dstEdge];

  const aggChoice = await ecmpLink(fiveTuple, salts[srcEdge], fabric.half);
  const srcAgg = `pod${srcPod}agg${aggChoice}`;

  if (srcPod === dstPod) {
    return [linkId(srcEdge, srcAgg), linkId(srcAgg, dstEdge)];
  }

  const coreChoice = await ecmpLink(fiveTuple, salts[srcAgg], fabric.half);
  const coreId = `core${aggChoice * fabric.half + coreChoice}`;
  const dstAgg = `pod${dstPod}agg${aggChoice}`;

  return [linkId(srcEdge, srcAgg), linkId(srcAgg, coreId), linkId(coreId, dstAgg), linkId(dstAgg, dstEdge)];
}

/** Ordered node path host->...->host for one flow, for the live animation.
 * Same routing as `route()` but returns the switch/host id sequence (including
 * both host endpoints) instead of the sorted link ids, so packets can be
 * tweened point-to-point. Empty when src/dst share an edge (no fabric hop). */
export async function routeNodes(
  fabric: Fabric,
  salts: Record<string, string>,
  fiveTuple: FiveTuple,
): Promise<string[]> {
  const srcHost = fabric.ipToHost[fiveTuple.src_ip];
  const dstHost = fabric.ipToHost[fiveTuple.dst_ip];
  const srcEdge = fabric.hostEdge[srcHost];
  const dstEdge = fabric.hostEdge[dstHost];
  if (srcEdge === dstEdge) return [];

  const srcPod = fabric.edgePod[srcEdge];
  const dstPod = fabric.edgePod[dstEdge];

  const aggChoice = await ecmpLink(fiveTuple, salts[srcEdge], fabric.half);
  const srcAgg = `pod${srcPod}agg${aggChoice}`;

  if (srcPod === dstPod) {
    return [srcHost, srcEdge, srcAgg, dstEdge, dstHost];
  }

  const coreChoice = await ecmpLink(fiveTuple, salts[srcAgg], fabric.half);
  const coreId = `core${aggChoice * fabric.half + coreChoice}`;
  const dstAgg = `pod${dstPod}agg${aggChoice}`;

  return [srcHost, srcEdge, srcAgg, coreId, dstAgg, dstEdge, dstHost];
}

/** Sorted link id joining two adjacent node ids (matches `fabric.linkIds`). */
export function segmentLinkId(a: string, b: string): string {
  return linkId(a, b);
}

/** North-south egress path host -> edge -> agg -> core -> WAN gateway, each hop
 * ECMP-hashed under the same salts. Models internet-bound traffic and animates
 * the super-spine tier. Empty if the fabric has no gateways. */
export async function egressRouteNodes(
  fabric: Fabric,
  salts: Record<string, string>,
  fiveTuple: FiveTuple,
): Promise<string[]> {
  if (fabric.gateways.length === 0) return [];
  const srcHost = fabric.ipToHost[fiveTuple.src_ip];
  const srcEdge = fabric.hostEdge[srcHost];
  const srcPod = fabric.edgePod[srcEdge];

  const aggChoice = await ecmpLink(fiveTuple, salts[srcEdge], fabric.half);
  const srcAgg = `pod${srcPod}agg${aggChoice}`;
  const coreChoice = await ecmpLink(fiveTuple, salts[srcAgg], fabric.half);
  const coreId = `core${aggChoice * fabric.half + coreChoice}`;
  const gwChoice = await ecmpLink(fiveTuple, salts[coreId], fabric.gateways.length);
  const gwId = fabric.gateways[gwChoice];

  return [srcHost, srcEdge, srcAgg, coreId, gwId];
}

/** Deterministic seeded PRNG (mulberry32): same seed -> identical stream, so the
 * traffic plan is byte-identical on every run and every salt source. */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export interface TrafficPlan {
  east: FiveTuple[]; // host<->host (some hosts are popular "servers")
  egress: FiveTuple[]; // host->WAN (north-south)
}

/** A realistic *non-uniform* traffic plan, fully deterministic (seeded): random
 * source hosts, destinations biased toward a small set of popular "server"
 * hosts (so a few edges run hot like a real DC), plus a heavy north-south egress
 * share to keep the WAN gateways busy. Fixed counts -> identical total traffic
 * every run; independent of salt source -> fair weak-vs-csprng comparison. */
export function buildTrafficPlan(
  fabric: Fabric,
  { seed = 0x5eed, east = 600, egress = 420, servers = 5 }: Partial<{ seed: number; east: number; egress: number; servers: number }> = {},
): TrafficPlan {
  const rng = mulberry32(seed);
  const hosts = fabric.hosts;
  const pick = () => hosts[Math.floor(rng() * hosts.length)];

  // A handful of popular destination hosts (the "servers" everyone talks to).
  const serverSet: string[] = [];
  while (serverSet.length < Math.min(servers, hosts.length)) {
    const h = pick();
    if (!serverSet.includes(h)) serverSet.push(h);
  }

  const eastFlows: FiveTuple[] = [];
  for (let i = 0; i < east; i++) {
    const src = pick();
    let dst = "";
    for (let tries = 0; tries < 8; tries++) {
      const candidate = rng() < 0.75 ? serverSet[Math.floor(rng() * serverSet.length)] : pick();
      if (fabric.hostEdge[candidate] !== fabric.hostEdge[src]) {
        dst = candidate;
        break;
      }
    }
    if (!dst) continue; // couldn't find a cross-edge dst; skip (deterministic)
    eastFlows.push({
      src_ip: fabric.hostToIp[src],
      dst_ip: fabric.hostToIp[dst],
      src_port: 40000 + (i % 4096),
      dst_port: 80,
      proto: 6,
    });
  }

  const egressFlows: FiveTuple[] = [];
  for (let i = 0; i < egress; i++) {
    egressFlows.push({
      src_ip: fabric.hostToIp[pick()],
      dst_ip: "203.0.113.9",
      src_port: 40000 + (i % 4096),
      dst_port: 443,
      proto: 6,
    });
  }

  return { east: eastFlows, egress: egressFlows };
}

/** Per-link flow counts over `flows`, ordered to match `fabric.linkIds`. */
export async function fabricLinkLoad(
  fabric: Fabric,
  salts: Record<string, string>,
  flows: FiveTuple[],
): Promise<number[]> {
  const counts: Record<string, number> = {};
  for (const id of fabric.linkIds) counts[id] = 0;
  for (const fiveTuple of flows) {
    const links = await route(fabric, salts, fiveTuple);
    for (const id of links) counts[id] += 1;
  }
  return fabric.linkIds.map((id) => counts[id]);
}

// ---------------------------------------------------------------------------
// Precision collision attacker (unified live stage).
//
// ONE compromised host floods a single victim with crafted flows (fixed dst,
// only src_port sweeps). It aims the whole flood at ONE deep fabric link, deep
// enough that ECMP has many equal-cost paths to spread across — the load-
// balancing tier the attack subverts. A first-hop edge uplink would be a poor
// target (only `half` parallel links, and the source's own uplinks congest
// regardless of salt); the interesting target is a core -> victim-agg link,
// where the fabric offers half^2 equal-cost core->agg paths into the victim's
// pod (k=6 -> 9 such paths).
//
// Every cross-pod flow to the victim traverses core -> victim-agg -> victim-edge
// (see route()). The path is fixed by two hashed choices: aggChoice (at the
// attacker's edge) and coreChoice (at its agg). The attacker crafts src_ports so
// both land on a chosen (a, c) => the flow rides core{a*half+c} -> victim-agg{a}
// = targetLink. It solves this against the salt it *believes* the switches use
// (`believedSalts`); packets route under the real salts (`realSalts`):
//   predictable salt (believed==real) => every crafted flow lands on (a,c) =>
//     the whole flood funnels onto the one target link and down to the victim
//     (red cone); the rest of the fabric is untouched. Targeted DoS.
//   unpredictable salt (believed!=real) => the same crafted ports hash to random
//     (aggChoice,coreChoice) => flows spray across all half^2 core->agg links =>
//     each carries ~1/half^2 => the flood dissolves into the background. Defeated.
// CSPRNG already breaks the precomputation; QRNG only adds attestable
// provenance, never extra mitigation (Experiment 4 null result). This is a
// SINGLE-attacker, single-target-link model by design — no botnet — matching
// the study's novelty framing (distinct from Crossfire/Coremelt).
// ---------------------------------------------------------------------------

export interface AttackForce {
  attacker: string; // the single compromised source host
  victim: string; // single target host
}

/** Deterministic attacker/victim pair on opposite ends of the fabric, in
 * different pods, so the crafted flood climbs edge -> agg -> core and back down
 * to the victim — maximally visible and forced through the core tier. */
export function attackForce(fabric: Fabric): AttackForce {
  return { attacker: fabric.hosts[fabric.hosts.length - 1], victim: fabric.hosts[0] };
}

/** The crafted 5-tuple: fixed attacker src, fixed victim dst, only src_port sweeps. */
function attackTuple(fabric: Fabric, attacker: string, victim: string, srcPort: number): FiveTuple {
  return {
    src_ip: fabric.hostToIp[attacker],
    dst_ip: fabric.hostToIp[victim],
    src_port: srcPort,
    dst_port: 80,
    proto: 6,
  };
}

export interface AttackPlan {
  routes: string[][]; // crafted flood paths, routed under the REAL salts
  targetLink: string; // the single deep core->victim-agg link aimed at
  targetAgg: string;
  collisionSetSize: number; // crafted flows in the flood
  scanned: number; // src_ports searched
  // Fraction of the flood that actually traverses targetLink under the REAL
  // salts — the concentration the attacker achieved. ~1.0 when the salt was
  // predictable (attack locks on), ~1/half^2 when it wasn't (scattered). This
  // is exact and background-independent, so it drives the locked/scattered
  // verdict without being fooled by ordinary traffic polarization.
  onTargetFraction: number;
}

/** Build the attacker's crafted flood aimed at one deep core->victim-agg link,
 * chosen by (targetAgg, targetCore). Scans src_port space keeping ports whose
 * (aggChoice, coreChoice) match the target *under the salt it believes*, then
 * routes each kept flow under the real salts so the animation shows where the
 * packets truly go. Predictable salt => believed==real => all converge;
 * unpredictable => they spray across the core tier. */
export async function craftAttackFlows(
  fabric: Fabric,
  realSalts: Record<string, string>,
  believedSalts: Record<string, string>,
  { attacker, victim }: AttackForce,
  {
    count = 200,
    window = 16000,
    targetAgg = 0,
    targetCore = 0,
  }: Partial<{ count: number; window: number; targetAgg: number; targetCore: number }> = {},
): Promise<AttackPlan> {
  const victimPod = fabric.edgePod[fabric.hostEdge[victim]];
  const targetAggId = `pod${victimPod}agg${targetAgg}`;
  const targetCoreId = `core${targetAgg * fabric.half + targetCore}`;
  const targetLink = linkId(targetCoreId, targetAggId);

  const attackerEdge = fabric.hostEdge[attacker];
  const attackerPod = fabric.edgePod[attackerEdge];
  const attackerAgg = `pod${attackerPod}agg${targetAgg}`;

  const routes: string[][] = [];
  let scanned = 0;
  for (let port = 40000; routes.length < count && port < 40000 + window; port++, scanned++) {
    const tuple = attackTuple(fabric, attacker, victim, port);
    const aggChoice = await ecmpLink(tuple, believedSalts[attackerEdge], fabric.half);
    if (aggChoice !== targetAgg) continue;
    const coreChoice = await ecmpLink(tuple, believedSalts[attackerAgg], fabric.half);
    if (coreChoice !== targetCore) continue;
    const nodes = await routeNodes(fabric, realSalts, tuple);
    if (nodes.length > 0) routes.push(nodes);
  }

  let onTarget = 0;
  for (const route of routes) {
    for (let i = 0; i < route.length - 1; i++) {
      if (linkId(route[i], route[i + 1]) === targetLink) {
        onTarget += 1;
        break;
      }
    }
  }

  return {
    routes,
    targetLink,
    targetAgg: targetAggId,
    collisionSetSize: routes.length,
    scanned,
    onTargetFraction: routes.length > 0 ? onTarget / routes.length : 0,
  };
}

/** A uniform background flow set: every distinct host pair, one flow each. */
export function uniformFlowSet(fabric: Fabric): FiveTuple[] {
  const flows: FiveTuple[] = [];
  for (const src of fabric.hosts) {
    for (const dst of fabric.hosts) {
      if (src === dst) continue;
      flows.push({
        src_ip: fabric.hostToIp[src],
        dst_ip: fabric.hostToIp[dst],
        src_port: 40000,
        dst_port: 80,
        proto: 6,
      });
    }
  }
  return flows;
}

File: TargetedDosColisionsAndRNGAngle/testbed/topology/fabric.py
"""Pure-data k=4 (Al-Fares) fat-tree model (plan-8). No Mininet/Ryu import --
this is the shared source of truth for the offline `polarization_check.py`
gate, the (heavier) live `FABRIC_MODE` controller, and the TS mirror
`web/lib/fabric.ts`. Reuses `hash_core.ecmp_link` unchanged (D8-parity).

Topology (k=4): `core_count = (k/2)**2` core switches, `k` pods, each pod has
`k/2` aggregation + `k/2` edge switches, each edge switch has `k/2` hosts --
4 core, 4 pods x (2 agg + 2 edge) = 16 switches, 20 switches total, 16 hosts.

Routing (D8-seed): a flow's egress at each switch is
`ecmp_link(five_tuple, salt, fanout)`. Only the **upward** hops are hashed
(edge->agg, agg->core), each using that switch's own per-switch salt (the
entropy knob). Downward (core->agg->edge->host) is deterministic by
destination pod/edge index, matching real ECMP fat-tree behaviour -- there is
nothing to hash once you know the destination subnet.

Honest-mechanism note (frozen in the plan): for k=4, `agg_per_pod == k // 2
== core_per_group`, so when an edge switch and its chosen aggregation switch
share the *same* salt (the fabric-wide-identical-prng case), both hops hash
the identical `sha256(five_tuple + salt)` digest against the *same* modulus
-- the two hop choices become perfectly correlated (`index2 == index1`
always), collapsing traffic onto a diagonal subset of core switches. This is
the polarization the offline check measures; it needs no attacker.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Sequence

from testbed.hash_core import ecmp_link
from testbed.salt.sources import salt_source
from testbed.types import FiveTuple


@dataclass(frozen=True)
class Fabric:
    k: int
    pods: int
    agg_per_pod: int
    edge_per_pod: int
    hosts_per_edge: int
    core_per_group: int
    core_count: int
    edge_switches: list[str] = field(default_factory=list)
    agg_switches: list[str] = field(default_factory=list)
    core_switches: list[str] = field(default_factory=list)
    all_switches: list[str] = field(default_factory=list)
    hosts: list[str] = field(default_factory=list)
    ip_to_host: dict[str, str] = field(default_factory=dict)
    host_to_ip: dict[str, str] = field(default_factory=dict)
    host_edge: dict[str, str] = field(default_factory=dict)
    edge_pod: dict[str, int] = field(default_factory=dict)
    agg_pod: dict[str, int] = field(default_factory=dict)
    agg_local_index: dict[str, int] = field(default_factory=dict)
    link_ids: list[str] = field(default_factory=list)


def _edge_name(pod: int, edge: int) -> str:
    return f"pod{pod}edge{edge}"


def _agg_name(pod: int, agg: int) -> str:
    return f"pod{pod}agg{agg}"


def _core_name(core: int) -> str:
    return f"core{core}"


def _host_name(pod: int, edge: int, host: int) -> str:
    return f"pod{pod}edge{edge}host{host}"


def _host_ip(pod: int, edge: int, host: int) -> str:
    return f"10.{pod}.{edge}.{host + 1}"


def _link_id(a: str, b: str) -> str:
    """Canonical (direction-independent) id for the single physical link
    between `a` and `b` -- a flow may traverse it upward or downward but it
    is the same link either way."""
    return "-".join(sorted((a, b)))


def build_fattree(k: int = 4) -> Fabric:
    """Canonical Al-Fares k-ary fat-tree: `pods = k`, `agg_per_pod = edge_per_pod
    = hosts_per_edge = core_per_group = k // 2`, `core_count = (k // 2) ** 2`."""
    half = k // 2
    edge_switches: list[str] = []
    agg_switches: list[str] = []
    core_switches: list[str] = [_core_name(c) for c in range(half * half)]
    all_switches: list[str] = list(core_switches)
    hosts: list[str] = []
    ip_to_host: dict[str, str] = {}
    host_to_ip: dict[str, str] = {}
    host_edge: dict[str, str] = {}
    edge_pod: dict[str, int] = {}
    agg_pod: dict[str, int] = {}
    agg_local_index: dict[str, int] = {}
    link_ids: list[str] = []

    for pod in range(k):
        for agg in range(half):
            agg_id = _agg_name(pod, agg)
            agg_switches.append(agg_id)
            all_switches.append(agg_id)
            agg_pod[agg_id] = pod
            agg_local_index[agg_id] = agg
        for edge in range(half):
            edge_id = _edge_name(pod, edge)
            edge_switches.append(edge_id)
            all_switches.append(edge_id)
            edge_pod[edge_id] = pod
            for agg in range(half):
                link_ids.append(_link_id(edge_id, _agg_name(pod, agg)))
            for host in range(half):
                host_id = _host_name(pod, edge, host)
                ip = _host_ip(pod, edge, host)
                hosts.append(host_id)
                ip_to_host[ip] = host_id
                host_to_ip[host_id] = ip
                host_edge[host_id] = edge_id
        for agg in range(half):
            agg_id = _agg_name(pod, agg)
            group_start = agg * half
            for core in range(group_start, group_start + half):
                link_ids.append(_link_id(agg_id, _core_name(core)))

    return Fabric(
        k=k,
        pods=k,
        agg_per_pod=half,
        edge_per_pod=half,
        hosts_per_edge=half,
        core_per_group=half,
        core_count=half * half,
        edge_switches=edge_switches,
        agg_switches=agg_switches,
        core_switches=core_switches,
        all_switches=all_switches,
        hosts=hosts,
        ip_to_host=ip_to_host,
        host_to_ip=host_to_ip,
        host_edge=host_edge,
        edge_pod=edge_pod,
        agg_pod=agg_pod,
        agg_local_index=agg_local_index,
        link_ids=link_ids,
    )


def fabric_salts(kind: str, fabric: Fabric) -> dict[str, bytes]:
    """Per-switch salt keyed by dpid/switch-id (D8-seed). `prng` -> one
    identical salt fabric-wide (models a same-image-same-seed deployment
    collapsing to a single seed everywhere -- OQ8-2); `csprng`/`qrng` ->
    an independent draw per switch."""
    if kind == "prng":
        shared = salt_source("prng").salt
        return {switch_id: shared for switch_id in fabric.all_switches}
    return {switch_id: salt_source(kind).salt for switch_id in fabric.all_switches}


def route(fabric: Fabric, salts: dict[str, bytes], five_tuple: FiveTuple) -> list[str]:
    """Walk src-edge -> (hashed) agg -> [same pod: straight to dst-edge |
    different pod: (hashed) core -> (deterministic) dst-agg -> (deterministic)
    dst-edge]. Returns the ordered link ids traversed. If src and dst share an
    edge switch, no hop is hashed (direct edge-host delivery) -- returns an
    empty list. Same-pod, different-edge traffic never detours through core
    (an aggregation switch reaches every edge switch in its own pod
    directly) -- mirrors real fat-tree behaviour and `next_hop()`'s per-switch
    decomposition exactly."""
    src_edge = fabric.host_edge[fabric.ip_to_host[five_tuple.src_ip]]
    dst_edge = fabric.host_edge[fabric.ip_to_host[five_tuple.dst_ip]]
    if src_edge == dst_edge:
        return []

    src_pod = fabric.edge_pod[src_edge]
    dst_pod = fabric.edge_pod[dst_edge]

    agg_choice = ecmp_link(five_tuple, salts[src_edge], fabric.agg_per_pod)
    src_agg = _agg_name(src_pod, agg_choice)

    if src_pod == dst_pod:
        return [_link_id(src_edge, src_agg), _link_id(src_agg, dst_edge)]

    core_choice = ecmp_link(five_tuple, salts[src_agg], fabric.core_per_group)
    core_id = _core_name(agg_choice * fabric.core_per_group + core_choice)

    dst_agg = _agg_name(dst_pod, agg_choice)

    return [
        _link_id(src_edge, src_agg),
        _link_id(src_agg, core_id),
        _link_id(core_id, dst_agg),
        _link_id(dst_agg, dst_edge),
    ]


def link_load(fabric: Fabric, salts: dict[str, bytes], flows: Sequence[FiveTuple]) -> list[int]:
    """Per-link flow counts over `flows`, ordered to match `fabric.link_ids`."""
    counts = {link_id: 0 for link_id in fabric.link_ids}
    for five_tuple in flows:
        for link_id in route(fabric, salts, five_tuple):
            counts[link_id] += 1
    return [counts[link_id] for link_id in fabric.link_ids]


def fabric_ports(fabric: Fabric) -> dict[str, dict[str, int]]:
    """Per-switch `{neighbor_id: port_no}`, port numbers assigned in the same
    link-creation order `fattree_topo.py`'s `build()` uses (inter-switch
    links from `fabric.link_ids`, then one host link per host in
    `fabric.hosts` order) so this matches Mininet's own port numbering."""
    ports: dict[str, dict[str, int]] = {switch_id: {} for switch_id in fabric.all_switches}
    counters: dict[str, int] = {switch_id: 0 for switch_id in fabric.all_switches}

    for link_id in fabric.link_ids:
        switch_a, switch_b = link_id.split("-")
        counters[switch_a] += 1
        ports[switch_a][switch_b] = counters[switch_a]
        counters[switch_b] += 1
        ports[switch_b][switch_a] = counters[switch_b]

    for host_id in fabric.hosts:
        edge_id = fabric.host_edge[host_id]
        counters[edge_id] += 1
        ports[edge_id][host_id] = counters[edge_id]

    return ports


def next_hop(fabric: Fabric, salts: dict[str, bytes], switch_id: str, five_tuple: FiveTuple) -> str:
    """The next-hop switch or host id for `five_tuple` arriving at `switch_id`
    (fabric-mode controller's per-switch decision -- the distributed
    decomposition of `route()`'s whole-path walk). Upward hops (edge->agg,
    agg->core) are hashed under `switch_id`'s own salt; downward hops are
    deterministic by destination pod/edge, matching `route()` exactly."""
    dst_host = fabric.ip_to_host[five_tuple.dst_ip]
    dst_edge = fabric.host_edge[dst_host]
    dst_pod = fabric.edge_pod[dst_edge]

    if switch_id in fabric.edge_switches:
        if switch_id == dst_edge:
            return dst_host
        pod = fabric.edge_pod[switch_id]
        agg_choice = ecmp_link(five_tuple, salts[switch_id], fabric.agg_per_pod)
        return _agg_name(pod, agg_choice)

    if switch_id in fabric.agg_switches:
        pod = fabric.agg_pod[switch_id]
        if pod == dst_pod:
            return dst_edge
        agg_local = fabric.agg_local_index[switch_id]
        core_choice = ecmp_link(five_tuple, salts[switch_id], fabric.core_per_group)
        return _core_name(agg_local * fabric.core_per_group + core_choice)

    # Core: deterministic to the agg in dst_pod belonging to this core's group.
    core_index = int(switch_id.removeprefix("core"))
    group = core_index // fabric.core_per_group
    return _agg_name(dst_pod, group)
